### PR TITLE
Enhance storage subsystem for coming features and fixes several bugs

### PR DESCRIPTION
**Pull request by jiangliu**
Enhance storage subsystem for coming features and fixes several bugs:
1) fix bugs related to compressed cache file and chunk digest validation etc
2) prepare for support non-continous compressed blob data
3) refine the way to decompress data

### CODE REVIEW

File: storage/src/cache/mod.rs [hunk 1]
```diff
@@ -210,13 +209,13 @@ pub trait BlobCache: Send + Sync {
     /// entry in the `chunks` array in corresponding order.
     ///
     /// This method returns success only if all requested data are successfully fetched.
-    fn read_chunks(
+    fn read_chunks_from_backend(
```
Comment (liubogithub) [inline storage/src/cache/mod.rs:212]: Nit: the above comment needs to be updated as well.

`Function read_chunks() returns one buffer `
to
`Function read_chunks_from_backend() returns one buffer `

File: storage/src/device.rs [hunk 2]
```diff
@@ -609,53 +618,43 @@ impl BlobIoVec {
...
+    pub fn is_empty(&self) -> bool {
+        self.bi_vec.is_empty()
+    }
+
+    /// Get size of pending IO data.
+    pub fn size(&self) -> u32 {
+        self.bi_size
+    }
+
+    /// Get `BlobIoChunk` associated with the specified `BlobIoDesc`.
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply storage/src/device.rs:570]: fixed

File: storage/src/device.rs [hunk 3]
```diff
@@ -1122,18 +1121,15 @@ impl FileReadWriteVolatile for BlobDeviceIoVec<'_> {
...
-        }
+        // BlobDevice::read_to() has validated that all IOs are against a single blob.
+        let index = self.iovec.blob_index();
+        let blobs = &self.dev.blobs.load();
 
-        Err(einval!("can not get blob index"))
+        if (index as usize) < blobs.len() {
+            blobs[index as usize].read(self.iovec, buffers)
+        } else {
+            Err(einval!("failed to get blob object for BlobIoVec"))
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply storage/src/device.rs:1067]: fixed

File: storage/src/device.rs [hunk 4]
```diff
@@ -515,11 +515,16 @@ impl BlobIoDesc {
     }
 
     /// Check whether the `other` BlobIoDesc is continuous to current one.
-    pub fn is_continuous(&self, prev: &BlobIoDesc) -> bool {
-        let offset = self.chunkinfo.compressed_offset();
-        let prev_size = prev.chunkinfo.compressed_size() as u64;
-        if let Some(prev_end) = prev.chunkinfo.compressed_offset().checked_add(prev_size) {
-            prev_end == offset && self.blob.blob_index() == prev.blob.blob_index()
+    pub fn is_continuous(&self, next: &BlobIoDesc, max_gap: u64) -> bool {
+        let prev_end = self.chunkinfo.compressed_offset() + self.chunkinfo.compressed_size() as u64;
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply storage/src/device.rs:483]: we are safe here:)

File: storage/src/cache/mod.rs [hunk 5]
```diff
@@ -51,6 +51,7 @@ pub const SINGLE_INFLIGHT_WAIT_TIMEOUT: u64 = 2000;
 
 struct BlobIoMergeState<'a, F: FnMut(BlobIoRange)> {
     cb: F,
+    // size of compressed daa
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply storage/src/cache/mod.rs:52]: fixed

File: storage/src/cache/mod.rs (hunk 1 above)
Comment (jiangliu) [reply storage/src/cache/mod.rs:212]: fixed

File: storage/src/cache/cachedfile.rs [hunk 6]
```diff
@@ -956,7 +956,7 @@ impl FileCacheEntry {
                 decoder.read_exact(buffer)?;
             }
         }
-        self.validate_chunk_data(chunk, buffer, false)?;
+        self.validate_chunk_data(chunk, buffer, self.need_validate)?;
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply storage/src/cache/cachedfile.rs:968]: fixed

File: storage/src/cache/mod.rs [hunk 7]
```diff
@@ -253,20 +253,11 @@ pub trait BlobCache: Send + Sync {
         chunk: &dyn BlobChunkInfo,
         buffer: &mut [u8],
         force_validation: bool,
-        raw_hook: Option<&dyn Fn(&[u8])>,
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply storage/src/cache/mod.rs:272]: fixed

**Review by liubogithub — APPROVED** [review]

**Pull request merged**